swabbie-redis/src/main/kotlin/com/netflix/spinnaker/swabbie/redis/RedisTaskTrackingRepository.kt

```diff
@@ -0,0 +1,139 @@
+  private val objectMapper: ObjectMapper,
+  private val clock: Clock
+) : TaskTrackingRepository {
+
+  private val log: Logger = LoggerFactory.getLogger(javaClass)
+
+  private val redisClientDelegate: RedisClientDelegate = redisClientSelector.primary("default")
+  private val SUBMITTED_TASKS_KEY = "{swabbie:submitted}"
+  private val TASK_STATUS_KEY = "{swabbie:taskstatus}"
+
+  override fun add(taskId: String, taskCompleteEventInfo: TaskCompleteEventInfo) {
+    if (taskCompleteEventInfo.submittedTimeMillis == null) {
+      taskCompleteEventInfo.submittedTimeMillis = clock.instant().toEpochMilli()
+    }
+
+    redisClientDelegate.withCommandsClient { client ->
+      client.hset(SUBMITTED_TASKS_KEY, taskId, objectMapper.writeValueAsString(taskCompleteEventInfo))
+      client.hset(TASK_STATUS_KEY, taskId, TaskState.IN_PROGRESS.toString())
+    }
+  }
+
+  override fun isInProgress(taskId: String): Boolean {
+    redisClientDelegate.run {
+      return this.withCommandsClient<Boolean> { client ->
+        client.hget(TASK_STATUS_KEY, taskId) == TaskState.IN_PROGRESS.toString()
+      }
+    }
+  }
+
+  override fun setSucceeded(taskId: String) {
+    redisClientDelegate.withCommandsClient { client ->
+      client.hset(TASK_STATUS_KEY, taskId, TaskState.SUCCEEDED.toString())
+    }
+  }
+
+  override fun setFailed(taskId: String) {
+    redisClientDelegate.withCommandsClient { client ->
+      client.hset(TASK_STATUS_KEY, taskId, TaskState.FAILED.toString())
+    }
+  }
+
+  override fun getInProgress(): Set<String> {
+    return getAll(TASK_STATUS_KEY)
+      .filterValues { it == TaskState.IN_PROGRESS.toString() }
+      .keys
+      .toSet()
+  }
+
+  override fun getFailed(): Set<String> {
+    return getAll(TASK_STATUS_KEY)
+      .filterValues { it == TaskState.FAILED.toString() }
+      .keys
+      .toSet()
+  }
+
+  override fun getSucceeded(): Set<String> {
+    return getAll(TASK_STATUS_KEY)
+      .filterValues { it == TaskState.SUCCEEDED.toString() }
+      .keys
+      .toSet()
+  }
+
+  private fun getAll(key: String): Map<String, String> {
+    return redisClientDelegate.run {
+      this.withCommandsClient<Map<String, String>> { client ->
+        client.hgetAll(key)
+      }
+    }
+  }
+
+  override fun getTaskDetail(taskId: String): TaskCompleteEventInfo? {
+    return redisClientDelegate.run {
+      this.withCommandsClient<String> { client ->
+        client.hget(SUBMITTED_TASKS_KEY, taskId)
+      }.let {
+        objectMapper.readValue(it, TaskCompleteEventInfo::class.java)
+      }
+    }
+  }
+
+  private fun getAllBefore(daysToGoBack: Int): Set<String> {
+    return getAll(SUBMITTED_TASKS_KEY)
+      .filterValues { value ->
+        objectMapper.readValue(value, TaskCompleteEventInfo::class.java).let { taskInfo ->
+          val xDaysAgo = clock.instant().minusMillis(TimeUnit.DAYS.toMillis(daysToGoBack.toLong())).toEpochMilli()
+          taskInfo.submittedTimeMillis!! < xDaysAgo
```

Comment:
`!!` is usually a code smell. Are you sure you want to hard-crash the process if the submitted time is null?

Reply:
We set this value when the object is added to the repository (see L52). Let me think about handling it in a different way..

Reply:
Fixed.